sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,157 @@
+  //  schema: i INTEGER, s STRING
+  //  filters: IsNotNull("i"), AlwaysTrue, Or(EqualTo("i", 0), StringStartsWith("s", "abc"))
+  //  predicates:
+  //    0: Array(IsNotNull("i"), AlwaysTrue, Or(EqualTo("i", 0), StringStartsWith("s", "abc")))
+  //    1: Array(AlwaysTrue, Or(EqualTo("i", 0), StringStartsWith("s", "abc")))
+  private val predicates: Array[Array[JsonPredicate]] = {
+    val groupedPredicates = Array.fill(schema.length)(Array.empty[JsonPredicate])
+    if (SQLConf.get.jsonFilterPushDown) {
+      val groupedByRefSet: Map[Set[String], JsonPredicate] = filters
+        // Group filters that have the same set of references. For example:
+        //   IsNotNull("i") -> Set("i"), AlwaysTrue -> Set(),
+        //   Or(EqualTo("i", 0), StringStartsWith("s", "abc")) -> Set("i", "s")
+        // By grouping filters we could avoid tracking their state of references in the
+        // current row separately.
+        .groupBy(_.references.toSet)
+        // Combine all filters from the same group by `And` because all filters should
+        // return `true` to do not skip a row. The result is compiled to a predicate.
+        .map { case (refSet, refsFilters) =>
+          (refSet, JsonPredicate(toPredicate(refsFilters), refSet.size))
+        }
+      // Apply predicates w/o references like `AlwaysTrue` and `AlwaysFalse` to all fields.
+      // We cannot set such predicates to a particular position because skipRow() can
+      // be invoked for any index due to unpredictable order of JSON fields in JSON records.
+      val withLiterals: Map[Set[String], JsonPredicate] = groupedByRefSet.map {
+        case (refSet, pred) if refSet.isEmpty =>
+          (schema.fields.map(_.name).toSet, pred.copy(totalRefs = 1))
+        case others => others
+      }
+      // Build a map where key is only one field and value is seq of predicates refer to the field
+      // "i" -> Seq(AlwaysTrue, IsNotNull("i"), Or(EqualTo("i", 0), StringStartsWith("s", "abc")))
+      // "s" -> Seq(AlwaysTrue, Or(EqualTo("i", 0), StringStartsWith("s", "abc")))
+      val groupedByFields: Map[String, Seq[(String, JsonPredicate)]] = withLiterals.toSeq
+        .flatMap { case (refSet, pred) => refSet.map((_, pred)) }
+        .groupBy(_._1)
+      // Build the final array by converting keys of `groupedByFields` to their
+      // indexes in the provided schema.
+      groupedByFields.foreach { case (fieldName, fieldPredicates) =>
+        val fieldIndex = schema.fieldIndex(fieldName)
+        groupedPredicates(fieldIndex) = fieldPredicates.map(_._2).toArray
+      }
+    }
+    groupedPredicates
+  }
+
+  /**
+   * Applies predicates (compiled filters) associated with the row field value
+   * at the position `index` only if other predicates dependencies are already
+   * set in the given row.
+   *
+   * Note: If the function returns `true`, `refCount` of some predicates can be non decremented.
```

Comment:
typo? `non` -> `not`?

Reply:
"non decremented" is also correct, isn't it? https://www.google.com/search?q=%22non+decremented%22&oq=%22non+decremented%22&aqs=chrome..69i57j0l2.2813j0j4&sourceid=chrome&ie=UTF-8

Reply:
Ya. I understand you may think like that. However, `non` is latin literally, isn't it? In addition, even in your search result, the top result seems to require hypen.